fdbserver/MasterProxyServer.actor.cpp

```diff
@@ -134,43 +144,66 @@ ACTOR Future<Void> getRate(UID myID, Reference<AsyncVar<ServerDBInfo>> db, int64
 	}
 }
 
+struct TransactionRateInfo {
+	double rate;
+	double limit;
+
+	TransactionRateInfo(double rate) : rate(rate), limit(0) {}
+
+	void reset(double elapsed) {
+		limit =
+		    std::min(0.0, limit) + std::min(rate * elapsed, SERVER_KNOBS->START_TRANSACTION_MAX_TRANSACTIONS_TO_START);
+	}
+
+	bool canStart(int64_t numAlreadyStarted) { return numAlreadyStarted < limit; }
+
+	void updateBudget(int64_t numStarted) { limit -= numStarted; }
+};
+
 ACTOR Future<Void> queueTransactionStartRequests(
-	std::priority_queue< std::pair<GetReadVersionRequest, int64_t>, std::vector< std::pair<GetReadVersionRequest, int64_t> > > *transactionQueue,
-	FutureStream<GetReadVersionRequest> readVersionRequests,
-	PromiseStream<Void> GRVTimer, double *lastGRVTime,
-	double *GRVBatchTime, FutureStream<double> replyTimes,
-	ProxyStats* stats) 
-{
+    std::priority_queue<std::pair<GetReadVersionRequest, int64_t>,
+                        std::vector<std::pair<GetReadVersionRequest, int64_t>>>* transactionQueue,
+    FutureStream<GetReadVersionRequest> readVersionRequests, PromiseStream<Void> GRVTimer, double* lastGRVTime,
+    double* GRVBatchTime, FutureStream<double> replyTimes, ProxyStats* stats, TransactionRateInfo* batchRateInfo) {
 	state int64_t counter = 0;
 	loop choose{
 		when(GetReadVersionRequest req = waitNext(readVersionRequests)) {
-			if (req.debugID.present())
+			if (req.debugID.present()) {
 				g_traceBatch.addEvent("TransactionDebug", req.debugID.get().first(), "MasterProxyServer.queueTransactionStartRequests.Before");
+			}
 
-			stats->txnStartIn += req.transactionCount;
-			if (req.priority() >= GetReadVersionRequest::PRIORITY_SYSTEM_IMMEDIATE)
+			if (req.priority() >= GetReadVersionRequest::PRIORITY_SYSTEM_IMMEDIATE) {
 				stats->txnSystemPriorityStartIn += req.transactionCount;
-			else if (req.priority() >= GetReadVersionRequest::PRIORITY_DEFAULT)
+			} else if (req.priority() >= GetReadVersionRequest::PRIORITY_DEFAULT) {
 				stats->txnDefaultPriorityStartIn += req.transactionCount;
-			else
+			} else {
+				// Return error for bath_priority GRV requests
```

Comment:
```suggestion
				// Return error for batch_priority GRV requests
```

Reply:
This was marked resolved but should still be applied

Reply:
Oops. Somehow I missed this change.